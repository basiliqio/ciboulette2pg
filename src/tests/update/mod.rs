use super::*;

mod query_params;
mod update_main_fails;
mod update_rel_fails;
mod update_rel_success;
mod update_success;

async fn test_update<'store>(
    pool: &mut sqlx::PgPool,
    query_end: &str,
    _test_name: &str,
    data: &BTreeMap<String, Vec<Uuid>>,
) -> Vec<sqlx::postgres::PgRow> {
    let ciboulette_store = gen_bag();
    let table_store = gen_table_store(&ciboulette_store);
    let parsed_url = Url::parse(format!("http://localhost{}", query_end).as_str()).unwrap();
    let first_person = data.get("peoples").unwrap().first().unwrap();
    let body = json!({
        "data": json!({
            "type": "peoples",
            "id": first_person,
            "attributes": json!({
                "first-name": "Sicnarf",
                "last-name": "Yor el",
            })
        })
    })
    .to_string();
    let body_opt = Some(body.as_str());
    const INTENTION: CibouletteIntention = CibouletteIntention::Update;

    let req_builder = CibouletteInboundRequestBuilder::new(INTENTION, &parsed_url, &body_opt);
    let request = req_builder.build(&ciboulette_store).unwrap();
    let ciboulette_request = CibouletteUpdateRequest::try_from(request).unwrap();
    let builder = Ciboulette2PostgresBuilder::gen_update(
        &ciboulette_store,
        &table_store,
        &ciboulette_request,
    )
    .unwrap();
    let (query, args) = builder.build().unwrap();

    let raw_rows: Vec<sqlx::postgres::PgRow> = sqlx::query_with(&query, args)
        .fetch_all(&mut pool.acquire().await.unwrap())
        .await
        .unwrap();
    snapshot_table(
        &mut *pool,
        "db_snapshot_update_main_while_testing_query_params",
        &["peoples", "people-article", "favorite_color"],
    )
    .await;
    raw_rows
}

async fn test_update_many_to_one<'store>(
    pool: &mut sqlx::PgPool,
    query_end: &str,
    _test_name: &str,
    data: &BTreeMap<String, Vec<Uuid>>,
) -> Vec<sqlx::postgres::PgRow> {
    let ciboulette_store = gen_bag();
    let table_store = gen_table_store(&ciboulette_store);
    let parsed_url = Url::parse(format!("http://localhost{}", query_end).as_str()).unwrap();
    let last_person = data.get("peoples").unwrap().last().unwrap();
    let body = json!({
        "data": json!({
            "type": "peoples",
            "id": last_person,
        })
    })
    .to_string();
    let body_opt = Some(body.as_str());
    const INTENTION: CibouletteIntention = CibouletteIntention::Update;

    let req_builder = CibouletteInboundRequestBuilder::new(INTENTION, &parsed_url, &body_opt);
    let request = req_builder.build(&ciboulette_store).unwrap();
    let ciboulette_request = CibouletteUpdateRequest::try_from(request).unwrap();
    let builder = Ciboulette2PostgresBuilder::gen_update(
        &ciboulette_store,
        &table_store,
        &ciboulette_request,
    )
    .unwrap();
    let (query, args) = builder.build().unwrap();

    let raw_rows: Vec<sqlx::postgres::PgRow> = sqlx::query_with(&query, args)
        .fetch_all(&mut pool.acquire().await.unwrap())
        .await
        .unwrap();
    snapshot_table(
        &mut *pool,
        "db_snapshot_update_rels_while_testing_query_params",
        &["comments"],
    )
    .await;
    raw_rows
}
