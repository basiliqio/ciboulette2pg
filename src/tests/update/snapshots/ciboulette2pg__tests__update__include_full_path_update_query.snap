---
source: src/tests/update/mod.rs
expression: "sqlformat::format(query.as_str(), &sqlformat::QueryParams::None,\n                  sqlformat::FormatOptions::default())"

---
WITH "cte_peoples_update_0" AS (
  UPDATE
    "public"."peoples"
  SET
    "first-name" = $1,
    "last-name" = $2
  WHERE
    "public"."peoples"."id" = $3 RETURNING *
),
"cte_peoples_data_1" AS (
  SELECT
    "cte_peoples_update_0"."id" :: TEXT AS "id",
    "cte_peoples_update_0"."id" AS "main_id",
    NULL :: TEXT AS "related_id",
    NULL :: TEXT AS "related_type",
    $4 :: TEXT AS "type",
    JSON_BUILD_OBJECT(
      $5,
      "cte_peoples_update_0"."age",
      $6,
      "cte_peoples_update_0"."first-name",
      $7,
      "cte_peoples_update_0"."gender",
      $8,
      "cte_peoples_update_0"."last-name",
      $9,
      "cte_peoples_update_0"."twitter"
    ) AS "data",
    "cte_peoples_update_0"."favorite_color" AS "rel_favorite_color"
  FROM
    "cte_peoples_update_0"
),
"cte_rel_articles_data_2" AS (
  SELECT
    "public"."articles"."id" :: TEXT AS "id",
    "public"."articles"."id" AS "main_id",
    "cte_peoples_data_1"."id" :: TEXT AS "related_id",
    $10 :: TEXT AS "related_type",
    $11 :: TEXT AS "type",
    JSON_BUILD_OBJECT(
      $12,
      "public"."articles"."body",
      $13,
      "public"."articles"."title"
    ) AS "data"
  FROM
    "public"."articles"
    INNER JOIN "public"."people-article" ON "public"."people-article"."article_id" = "public"."articles"."id"
    INNER JOIN "cte_peoples_data_1" ON "cte_peoples_data_1"."main_id" = "public"."people-article"."people_id"
) (
  SELECT
    "cte_peoples_data_1"."id",
    "cte_peoples_data_1"."type",
    "cte_peoples_data_1"."data",
    "cte_peoples_data_1"."related_type",
    "cte_peoples_data_1"."related_id"
  FROM
    "cte_peoples_data_1"
)
UNION
ALL (
  SELECT
    "cte_rel_articles_data_2"."id",
    "cte_rel_articles_data_2"."type",
    "cte_rel_articles_data_2"."data",
    "cte_rel_articles_data_2"."related_type",
    "cte_rel_articles_data_2"."related_id"
  FROM
    "cte_rel_articles_data_2"
);
